src/Microsoft.Diagnostics.Monitoring.WebApi/Metrics/JsonCounterLogger.cs

```diff
@@ -3,21 +3,41 @@
 // See the LICENSE file in the project root for more information.
 
 using Microsoft.Diagnostics.Monitoring.EventPipe;
+using System;
+using System.Buffers;
 using System.IO;
 using System.Text.Json;
+using System.Threading.Tasks;
 
 namespace Microsoft.Diagnostics.Monitoring.WebApi
 {
     internal sealed class JsonCounterLogger : StreamingCounterLogger
     {
-        public JsonCounterLogger(Stream stream) : base(stream)
+        // The default metrics providers will produce about 470 bytes of data
+        // at most per counter value. Set the buffer to be slightly larger to
+        // avoid reallocation.
+        private const int InitialBufferCapacity = 500;
+
+        private static readonly ReadOnlyMemory<byte> JsonSequenceRecordSeparator =
+            new byte[] { StreamingLogger.JsonSequenceRecordSeparator };
+        private static readonly ReadOnlyMemory<byte> NewLineSeparator =
+            new byte[] { (byte)'\n' };
+
+        private readonly ArrayBufferWriter<byte> _bufferWriter;
+        private readonly Stream _stream;
+
+        public JsonCounterLogger(Stream stream)
         {
+            _stream = stream;
+            _bufferWriter = new(InitialBufferCapacity);
         }
 
-        protected override void SerializeCounter(Stream stream, ICounterPayload counter)
+        protected override async Task SerializeAsync(ICounterPayload counter)
         {
-            stream.WriteByte(StreamingLogger.JsonSequenceRecordSeparator);
-            using (var writer = new Utf8JsonWriter(stream, new JsonWriterOptions { Indented = false }))
```

Comment:
Related topic: https://github.com/dotnet/runtime/issues/30401

Note the approach in that issue would force some aggressive flushing on our part so I'm not suggesting we use it.